icu4c/source/tools/genrb/reslist.cpp

```diff
@@ -1338,13 +1341,39 @@ compareKeyOldpos(const void * /*context*/, const void *l, const void *r) {
     return compareInt32(((const KeyMapEntry *)l)->oldpos, ((const KeyMapEntry *)r)->oldpos);
 }
 
+void SResource::collectKeys(std::function<void(int32_t)> collector) const {
+    collector(fKey);
+}
+
+void ContainerResource::collectKeys(std::function<void(int32_t)> collector) const {
+    collector(fKey);
+    for (SResource* curr = fFirst; curr != NULL; curr = curr->fNext) {
+        curr->collectKeys(collector);
+    }
+}
+
 void
 SRBRoot::compactKeys(UErrorCode &errorCode) {
     KeyMapEntry *map;
     char *keys;
     int32_t i;
-    int32_t keysCount = fUsePoolBundle->fKeysCount + fKeysCount;
-    if (U_FAILURE(errorCode) || fKeysCount == 0 || fKeyMap != NULL) {
+
+    // Except for pool bundles, keys might not be used.
+    // Do not add unused keys to the final bundle.
+    std::set<int32_t> keysInUse;
+    if (fIsPoolBundle) {
+        fKeysMapCount = fKeysTotalCount;
+    } else {
+        fRoot->collectKeys([&keysInUse](int32_t key) {
+            if (key >= 0) {
+                keysInUse.insert(key);
+            }
+        });
+        fKeysMapCount = keysInUse.size();
+    }
+
+    int32_t keysCount = fUsePoolBundle->fKeysCount + fKeysMapCount;
+    if (U_FAILURE(errorCode) || fKeysTotalCount == 0 || fKeyMap != NULL) {
```

Comment:
Should fKeysTotalCount really be fKeysMapCount?
That is, if no keys at all are used after filtering, we can stop, right?

See also my related question in the .h file.